### PR TITLE
AIAT-129 / Time tracking (socket.io implementation)

### PR DESCRIPTION
**Pull request by jcarstairs-scottlogic**
[Ticket AIAT-129](https://scottlogic.atlassian.net/browse/AIAT-129?atlOrigin=eyJpIjoiZGEwZTJmMGQ3Y2EyNDY1ZThmNGY4ZDhlYzkwYzBjYWMiLCJwIjoiaiJ9)

Uses [socket.io](https://socket.io) to emit a stream of events from the tree test study frontend to the server.

Records these events:

- page load
- activate node
- open node (when you expand a parent node to show its children: sometimes coincides with activating it, but not always)
- close node (when you fold in a parent node to hide its children: sometimes coincides with activating it, but not always)
- task changed
- submit response
- window visibility changed

All events include **at least** these data:

- time sent
- time received
- tree test ID
- event type

The client will always try and make the server consistent, even if the connection drops. I've used the [acknowledgements and timeouts features of socket.io](https://socket.io/docs/v4/emitting-events/#with-timeout) to achieve this.

I couldn't exactly achieve one of the AC: to emit an event when the user closes the browser. The [Window beforeunload event](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event) is close, but according to MDN it is unreliable. Instead I used the [Document visibilitychange event](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event). As well as firing when the user closes the tab or window, it also fires when the user changes tab or minimises the window. I think this is OK, since I think the intention was to identify when the user is no longer actively working on the task, and if the user changes tab or minimises the window, they probably aren't actively working on the task any more. One weakness I've found of this approach is that it doesn't fire if you kill the browser process from Windows Task Manager, and I'm guessing it won't fire if the browser crashes, so we won't always be able to correctly infer when the user was working on the task.

### CODE REVIEW

File: js/treetest.js [hunk 1]
```diff
@@ -115,8 +130,80 @@ var tasks = {
...
 }
+
+var socket = {
+	_socket: null,
+
+	connect: function() {
+		this._socket = io();
+		this._socket.on('connect', () => {
+			console.debug('Socket connected');
+			this._socket.on('disconnect', () => {
```
Comment (pthanvi) [inline js/treetest.js]: Based on socket.io documentation, we should refrain from registering in the `connect` handler since it will result in a new  registration every time the instance re-connects

https://socket.io/docs/v4/client-socket-instance/#connect

Comment (pthanvi) [inline js/treetest.js]: Can we also log the reason for socket disconnection, this will help us to troubleshoot why connection was dropped.

File: js/treetest.js [hunk 2]
```diff
@@ -115,8 +130,80 @@ var tasks = {
...
+	emitWindowVisibilityChangedEvent: function(newVisibilityState) {
+		const data = { new_visibility_state: newVisibilityState };
+		this._emitEvent(data, 'window visibility changed');
+	},
+
+	_emitEvent: function(data, eventType) {
+		const json = JSON.stringify({
+			id: crypto.randomUUID(),
+			timestamp: new Date().toISOString(),
+			tree_test_id: document.split('/').pop(),
```
Comment (pthanvi) [inline js/treetest.js]: The id in the URL is in fact the response id. Can we change the variable name and the field in the schema to reflect this. 

Ideally we should also store the study id, but we can derive this using the response id since each response is mapped to the study, so we should be good.

Comment (pthanvi) [inline js/treetest.js]: Can we store this somewhere to be reused rather than deriving it every time

File: models/event.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+var mongoose = require('mongoose');
+
+var Event = new mongoose.Schema({
+	id: String,
+	type: String,
+	iso_timestamp_sent: String,
+	iso_timestamp_received: String,
+	tree_test_id: String,
```
Comment (pthanvi) [inline models/event.js]: As mentioned elsewhere, this should be response id

File: models/event.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+var mongoose = require('mongoose');
+
+var Event = new mongoose.Schema({
+	id: String,
+	type: String,
+	iso_timestamp_sent: String,
```
Comment (pthanvi) [inline models/event.js]: Looks like MongoDB does not have any specific naming convention for field names, but can we use camel case here instead so that it is consistent with the field names used in other collections.

File: views/treetest/view.ejs [hunk 5]
```diff
@@ -8,8 +8,9 @@
 		<!-- custom css -->
 		<link rel="stylesheet" type="text/css" href="/css/treetest.css" />
 		<title>Kort Treetest</title>
-		<script type="text/javascript" src="/js/treetest.js"/></script>
+		<script type="text/javascript" src="/js/treetest.js"></script>
 		<script type="text/javascript" src="/js/treetest-view.js"></script>
+		<script type="text/javascript" src="/socket.io/socket.io.js"></script>
```
Comment (pthanvi) [inline views/treetest/view.ejs:13]: Since `view.ejs` is also used when previewing, I think we will end up capturing events even in preview mode which we do not want.

File: js/treetest.js [hunk 6]
```diff
@@ -140,4 +227,53 @@ function setup(input_tasks,input_tree,input_selectableParents,input_closeSibling
...
+function bindEmitOnCloseNode() {
+	// Fires when the user close a parent node to hideits children, either by
+	// activating the parent node itself, or by activating the button next to it
+	$('#tree').on('close_node.jstree', function (_, { node }) {
+		socket.emitCloseNodeEvent(node);
+	});
+}
+
+function bindEmitOnTaskChanged() {
+	tasks.onSet.push(number => socket.emitTaskChangedEvent(number));
```
Comment (pthanvi) [inline js/treetest.js]: It doesn't look like we are binding anything here. If I understand correctly, we are adding a lambda to a list (`onSet`) and when the user moves to a different task we loop through the list and call the lambda passing it the task number.

It feels like a very convoluted way of doing things here. Why can't we directly emit the `TaskChanged` event rather than looping and calling the lambda?

File: js/treetest.js [hunk 7]
```diff
@@ -140,4 +227,53 @@ function setup(input_tasks,input_tree,input_selectableParents,input_closeSibling
...
+}
+
+function bindEmitOnSubmitResponse() {
+	tasks.onSubmit.push(answers => socket.emitSubmitResponseEvent(answers));
```
Comment (pthanvi) [inline js/treetest.js]: Same comment as mentioned for `TaskChanged` event

File: js/treetest.js [hunk 8]
```diff
@@ -30,34 +32,44 @@ function initializeTreeViewObject(treeStructure){
...
+		tasks.answers[tasks.idx] = setHistory(data.node);
+		enableButton('#nextTaskButton');
 	});
 	$('#tree').on("deselect_node.jstree", function (e, data) {
 	  disableButton('#nextTaskButton');
 	});
 }
+
 function resetTree(){
+	const { emitCloseNodeEvent } = socket;
```
Comment (pthanvi) [inline js/treetest.js]: If I understand correctly, we are ensuring that `CloseNode` event is not raised here when we call close_all. Can you add a comment around this please.

File: js/treetest.js [hunk 9]
```diff
@@ -115,8 +130,80 @@ var tasks = {
...
+
+	emitOpenNodeEvent: function(node) {
+		const data = {
+			node: $('#tree').jstree().get_path(node),
+			current_task_index: tasks.idx,
+		};
+		this._emitEvent(data, 'open node');
+	},
+
+	emitCloseNodeEvent: function(node) {
```
Comment (pthanvi) [inline js/treetest.js:179]: There is a setting `Hide Siblings` in tree test which I think automatically collapses open nodes. Will that cause an event capture as well, not sure if we would want that. 

I could not test it since things are not working for me

File: js/treetest.js [hunk 10]
```diff
@@ -115,8 +130,80 @@ var tasks = {
...
+
+	_emitEvent: function(data, eventType) {
+		const json = JSON.stringify({
+			id: crypto.randomUUID(),
+			timestamp: new Date().toISOString(),
+			tree_test_id: document.split('/').pop(),
+			...data,
+		});
+		const emitUntilAcknowledged = () => this._socket
+			.timeout(2000)
```
Comment (pthanvi) [inline js/treetest.js:208]: Based on the documentation, any pending event will be lost if the user refreshes the tab
https://socket.io/docs/v4/delivery-guarantees

File: views/treetest/view.ejs (hunk 5 above)
Comment (jcarstairs-scottlogic) [reply views/treetest/view.ejs:13]: Good point! What I've done to fix this is get the response ID from the `<input id="resid">` on line 25. When previewing, this will be set to `preview`, so preview runs are identifiable in the database and can be filtered out

File: js/treetest.js (hunk 6 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js]: You could just call `emitTaskChangedEvent()` directly from `tasks.set()`. I just thought that would be mixing concerns -- in principle, I think the tasks logic shouldn't have to worry about the event emission logic. By using a callback, we can keep the concerns separated. In practice maybe it's not worth the extra verbosity in this case: what do you think?

(Same comment for the `emitSubmitResponseEvent` bit)

File: js/treetest.js (hunk 9 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:179]: Yes. Given `Hide siblings` is enabled, if you open one node, and this automatically causes another node to close, that means we get a 'close node' event.

It's already the case that you can sometimes get more than one event emitted per user action. This wouldn't be the only such case. For example, if you open a parent node by activating it, you get both an 'open node' event and an 'activate node' event.

@rvango doesn't seem to like this behaviour either, I think he's working on explaining why that is and what he would like to happen instead

File: js/treetest.js (hunk 10 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:208]: Indeed. I think that is the limits of how much SocketIO does to guarantee delivery. I think if we want better guarantees, we will have to store events locally ourselves, e.g. with local storage or session storage, and using that storage to guarantee delivery would be a significant amount of extra complexity I think

File: js/treetest.js [hunk 11]
```diff
@@ -91,6 +102,7 @@ var tasks = {
 			updateProgressBar();
 		} else {
 			$('#hiddenResults').val(JSON.stringify(tasks.answers));
+			this.onSubmit.forEach(callback => callback(tasks.answers));
```
Comment (rvango) [inline js/treetest.js]: I find it strange that we have a list of callbacks when we submit a response. I would have expected only 1

File: js/treetest.js [hunk 12]
```diff
@@ -109,14 +121,97 @@ var tasks = {
 		this.idx = number;
 		$('#taskDesc').html(this.list[number]);
 		$('#taskNum').html("Task "+(number+1)+" of "+this.list.length);
+
+		// Resetting the tree between tasks results in all open nodes closing, but
+		// we don't want to register these as close node events, because they aren't
+		// meaningfully user actions. So we pause listening for close node events
```
Comment (rvango) [inline js/treetest.js:127]: Is that the case where a node hierarchy get auto-closed when we expend another node?
I am struggling to understand what is happening here. Can you provide more details on this use case please?

File: js/treetest.js (hunk 7 above)
Comment (rvango) [reply js/treetest.js]: The intent is not obvious, I feel.
Would you be able to explain why using a Set() if we only have 1 callback? Do you expect more? If so, I rather keep it simple, and turn it into a set if we have to.

File: js/treetest.js (hunk 6 above)
Comment (rvango) [reply js/treetest.js]: Ok to wrap it into a function, but not convinced we need to use a Set as collection for our callback. 
I feel I possibly don't understand the reason for needing it to begin with. Can we just call the callback directly?

File: server/socket.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+		logger.info(`Socket connected: ${socket.id}`);
+		socket.on('disconnect', (reason) => {
+			logger.info(`Socket disconnected: ${socket.id}. Reason: ${reason}`);
+		});
+
+		socket.on('page load', saveEventToDb('page load'));
+		socket.on('activate node', saveEventToDb('activate node'));
+		socket.on('open node', saveEventToDb('open node'));
+		socket.on('close node', saveEventToDb('close node'));
+		socket.on('task changed', saveEventToDb('task changed'));
```
Comment (rvango) [inline server/socket.js:18]: Could we possibly get rid of this event? We should never ask more than one task against a tree.

File: js/treetest.js [hunk 14]
```diff
@@ -109,14 +121,97 @@ var tasks = {
...
+		const { emitCloseNodeEvent } = socket;
```
Comment (rvango) [inline js/treetest.js:128]: It is not clear why a `socket` is assigned into `emitCloseNodeEvent`
Should only one method from it be temporary saved?

Can you think of a better way to do **pause**/**do an action**/**unpause**?
Possibly a separate method?

File: js/treetest.js [hunk 15]
```diff
@@ -109,14 +121,97 @@ var tasks = {
...
 			expandToNode(copyOfHistory);
 		}
-	}
+		this.onSet.forEach(callback => callback(number));
+	},
+	onSet: [],
+	onSubmit: [],
+}
+
+var socket = {
```
Comment (rvango) [inline js/treetest.js:146]: It would benefit from be stored into its own file please.

File: js/treetest.js (hunk 9 above)
Comment (rvango) [reply js/treetest.js:179]: To keep it simple, I would have like to have `open` and `activated` to be the same event, and possibly be renamed as `selected`.

I might do that in a follow-up ticket because I reckon that might be a bit more tricky using the generic `check_callback()` that gets called for any events

```js
$('#your_tree_id').jstree({
    'core': {
        'check_callback': function(operation, node, node_parent, node_position, more) { 
          ...
    }
```

File: js/treetest.js (hunk 10 above)
Comment (rvango) [reply js/treetest.js:208]: Does this covers the following AC?
```
Given that a user is connected via real-time communication, 
When the connection drops, 
Then the client temporarily stores the events locally and retries sending upon reconnection, ensuring no events are lost.
```

File: js/treetest.js [hunk 16]
```diff
@@ -140,4 +235,53 @@ function setup(input_tasks,input_tree,input_selectableParents,input_closeSibling
...
+function bindEmitOnActivateNode() {
+	// Fires when the user activates a node (by clicking on it)
+	$('#tree').on('activate_node.jstree', function (_, { node }) {
+		socket.emitActivateNodeEvent(node);
+	});
+}
+
+function bindEmitOnOpenNode() {
+	// Fires when the user opens a parent node to expose its children, either by
+	// activating the parent node itself, or by activating the button next to it
```
Comment (rvango) [inline js/treetest.js:260]: Does it mean that `bindEmitOnOpenNode()` does cover `bindEmitOnActivateNode()`?

Is it then its superset?

File: js/treetest.js [hunk 17]
```diff
@@ -140,4 +235,53 @@ function setup(input_tasks,input_tree,input_selectableParents,input_closeSibling
...
+	$('#tree').on('open_node.jstree', function (_, { node }) {
+		socket.emitOpenNodeEvent(node);
+	});
+}
+
+function bindEmitOnCloseNode() {
```
Comment (rvango) [inline js/treetest.js:266]: I am wondering if we need that event. I am basically only interested by activating/expanding.

File: js/treetest.js (hunk 6 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js]: I think you're both agreeing it's too verbose in this case. I'll call `emitTaskChangedEvent()` directly from `tasks.set()`

File: js/treetest.js (hunk 10 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:208]: I interpreted 'when the connection drops' to mean the client is still running, but can't communicate with the server. These kinds of cases, SocketIO handles for free on the client:

- The user's laptop gets disconnected from their WiFi network  :heavy_check_mark:
- A random temporary network error in a data centre in Manchester somewhere prevents data getting through for a few seconds  :heavy_check_mark:
- The user's phone loses 3G signal :heavy_check_mark:

But SocketIO doesn't use any persistent storage, so it can't cover cases like these:

- The user refreshes the tab :x:
- The browser crashes :x:
- The user kills the browser process :x:
- The user drops their phone and it smashes :x:

To handle these cases, we would need storage which persists across user sessions, and we would need to implement and manage this ourselves. I think that would be quite a lot of extra complexity

Comment (rvango) [reply js/treetest.js:208]: That seems suitable to me. 👍

File: js/treetest.js (hunk 9 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:179]: Me and Remi had a chat about this, and we think it's fine to emit all the events. Later on, we can use the list of events to count how many actions the user took, by the following algorithm:

```mermaid
flowchart TB
    list[/input: EVENTS/]
    filter[filter EVENTS: 'open' or 'activate']
    sort[sort EVENTS: oldest to newest]
    zero[set COUNT: 0]
    next{"get next(EVENTS) as E"}
    equal{"node(E) = N?"}
    set["set N: node(E)"]
    increment[increment COUNT]
    output[/output COUNT/]
    list --> filter
    filter --> sort
    sort --> zero
    zero --> next
    next -->|E exists| equal
    equal -->|yes| next
    equal -->|no| set
    set --> increment
    increment --> next
    next ----->|no next event| output
```

File: js/treetest.js (hunk 14 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:128]: This is [destructuring assignment](https://javascript.info/destructuring-assignment), it's a common pattern in JavaScript. This line is equivalent to:

```js
const emitCloseNodeEvent = socket.emitCloseNodeEvent;
```

File: js/treetest.js (hunk 17 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:266]: After our chat, I think you're right, we don't need it, but it doesn't do any harm

File: js/treetest.js (hunk 16 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:260]: Activating a node sometimes opens it, too. But not always. And there is another way to open a node besides activating it. So you can activate a node without opening it, and you can open a node without activating it. So emitting an 'open node' event is not redundant

File: js/treetest.js (hunk 14 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:128]: Does that help? Or have I not understood the problem?

File: server/socket.js (hunk 13 above)
Comment (jcarstairs-scottlogic) [reply server/socket.js:18]: After our chat, I think you're right, we don't need it, but it doesn't do any harm

File: js/treetest.js (hunk 12 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:127]: When the task changes, the tree gets reset. This includes at page load when the active task is set to the first task. So even if there is only one task, this will still happen once. It is important not to accidentally emit 'close node' events just because the tree is being set up.

File: js/treetest.js (hunk 8 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js]: See [commit 41d8d9](https://github.com/ScottLogic/kort/pull/7/commits/41d8d983269ef5168c2d4b1c02260682db0b79aa)

File: views/treetest/view.ejs (hunk 5 above)
Comment (jcarstairs-scottlogic) [reply views/treetest/view.ejs:13]: See [commit b70280](https://github.com/ScottLogic/kort/pull/7/commits/b70280a6111a8655d46932c118cadea5992ed768)

File: js/treetest.js (hunk 6 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js]: Update: I found I could fire a custom Window Event to do the same thing in a nice idiomatic way. Hopefully this is easier to understand, and also keeps concerns separate. See [commit 3e3123](https://github.com/ScottLogic/kort/pull/7/commits/3e312303578207b6d6762e3b32ca7da278aee405)

File: js/treetest.js (hunk 15 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:146]: There are pros and cons of pulling this out to its own file.

Pros:

- Each file does one thing
- Each file is small

Cons:

- There is no syntactical link between the files, which will break tools which depend on static analysis, like symbol renaming, goto definition, type checking etc
- The client has to make an extra GET request for the extra file, which potentially means a slower load time

Ideally, we would have some kind of build step, so you could write files as separate modules and they would be concatenated into a single file at build time. But the project isn't currently set up to do that, which is annoying. I thought in this case, the cons outweighed the pros. The fact you can't do static analysis is quite a big disadvantage, I think.

**Review by rvango — APPROVED** [review]
Great job! :)

File: js/treetest.js (hunk 6 above)
Comment (pthanvi) [reply js/treetest.js]: This looks much cleaner, however I think there is a bug with the custom event. Both task changed and submit events does not capture any data.

File: models/event.js [hunk 18]
```diff
@@ -0,0 +1,12 @@
+var mongoose = require('mongoose');
+
+var Event = new mongoose.Schema({
+	id: String,
```
Comment (pthanvi) [inline models/event.js]: Minor - I think MongoDB uses a auto generated `_id` field for documents to ensure uniqueness. Since we are already providing our own unique id, `_id` field seems unnecessary. 

Is there a way to remove the default `_id` field? Don't bother if it is too much hassle.

Comment (jcarstairs-scottlogic) [reply models/event.js]: I think it can't be removed ([MongoDB docs](https://www.mongodb.com/docs/manual/core/document/#the-_id-field)). I could set `_id` to be the same value as our own `id`, but I don't see that it makes a difference

File: js/treetest.js (hunk 6 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js]: I've double checked, and I definitely get both events. Are you sure your branch is up to date with the remote? Are you running in Docker?

Comment (jcarstairs-scottlogic) [reply js/treetest.js]: Good spot, should be fixed now

File: js/treetest.js [hunk 19]
```diff
@@ -109,14 +121,96 @@ var tasks = {
 		this.idx = number;
 		$('#taskDesc').html(this.list[number]);
 		$('#taskNum').html("Task "+(number+1)+" of "+this.list.length);
+
+		// Resetting the tree between tasks results in all open nodes closing, but
+		// we don't want to register these as close node events, because they aren't
+		// meaningfully user actions. So we pause listening for close node events
+		const { emitCloseNodeEvent } = socket;
```
Comment (rvango) [inline js/treetest.js:128]: Very minor comment, but this way of pausing/unpausing requires a comment to understand. It might be clearer if it was wrapped neatly into an event handler manager to pause/unpause.

Something like that (not tested!):

```javascript
const eventHandlerManager = {
    handlers: new Map(),

    // Function to pause an event
    pauseEvent(socket, eventName) {
        if (this.handlers.has(eventName)) {
            console.log(`Event ${eventName} is already paused.`);
            return;
        }
        // Store the current event handler
        const originalHandler = socket[eventName];
        this.handlers.set(eventName, originalHandler);
        // Replace the event handler with a noop function
        socket[eventName] = () => {};
    },

    // Function to unpause an event
    unpauseEvent(socket, eventName) {
        if (!this.handlers.has(eventName)) {
            console.log(`Event ${eventName} is not paused or does not exist.`);
            return;
        }
        // Restore the original event handler
        socket[eventName] = this.handlers.get(eventName);
        this.handlers.delete(eventName);
    }
};

// Example usage:
const socket = {
    emitCloseNodeEvent: () => { console.log("Node closed"); }
};

// Pausing the close node event
eventHandlerManager.pauseEvent(socket, 'emitCloseNodeEvent');
console.log("Event paused");

// Example of trying to emit event while paused
socket.emitCloseNodeEvent();  // No output, as the handler is a noop

// Unpausing the close node event
eventHandlerManager.unpauseEvent(socket, 'emitCloseNodeEvent');
console.log("Event unpaused");

// Emitting the event after unpausing
socket.emitCloseNodeEvent();  // Outputs: "Node closed"
```

I often find code intent in JS hard to follow because I am not very familiar with the language. It would help me if behaviours or concerns were wrapped into object. 
One downside is that's soooo verbose... 👎

File: js/treetest.js [hunk 20]
```diff
@@ -109,14 +121,96 @@ var tasks = {
...
+		socket.emitCloseNodeEvent = emitCloseNodeEvent;
+
 		if(this.answers[this.idx].length > 0){
 			//need to pass a copy of node path to expandToNode (or else it alters tasks.answers)
 			var copyOfHistory = $.extend(true, [], this.answers[this.idx]);
 			expandToNode(copyOfHistory);
 		}
-	}
+
+		window.dispatchEvent(new CustomEvent('taskchanged', { detail: { newTaskIndex: number } }));
```
Comment (rvango) [inline js/treetest.js:142]: That's really a neat way to bind event natively. Great stuff!

File: models/event.js (hunk 18 above)
Comment (rvango) [reply models/event.js]: I would agree with Pallavi. If we could avoid creating another ID, it might be beneficial. Could we use our generated `id` to be used as `_id` so we don't have a duplicated field serving the same purpose?

**Review by rvango — APPROVED** [review]
Nice work Joe, I would like to have a second approval from @pthanvi also :)

File: js/treetest.js (hunk 19 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:128]: I agree that it's a better solution and also verbose. If we were doing more of this pausing/re-starting behaviour, I think it would be worth it -- but my judgement is that it's overkill right now

File: models/event.js (hunk 18 above)
Comment (jcarstairs-scottlogic) [reply models/event.js]: MongoDB docs (linked above) says that `_id` is always present. I can't find any hint that there are any exceptions or ways to change this

I also don't see what the benefit is of removing the `_id`

Comment (rvango) [reply models/event.js]: Not removing the `_id` but inject our UUID into it, instead of auto generating it as per the doc:
image

Comment (pthanvi) [reply models/event.js]: Just checked, if we provide our own `_id` field MongoDB does not create one. So we just need to change our field name to `_id`.
https://www.mongodb.com/docs/manual/core/document/#the-_id-field

File: js/treetest.js (hunk 19 above)
Comment (rvango) [reply js/treetest.js:128]: Readability is never overkill ;)

File: models/event.js (hunk 18 above)
Comment (jcarstairs-scottlogic) [reply models/event.js]: OK, do you want both `_id` and `id` or to replace `id` with `_id`? I still don't quite understand what the problem is

File: js/treetest.js (hunk 6 above)
Comment (pthanvi) [reply js/treetest.js]: It works now :)

Sorry to be a pain, but can you increment task number by 1 when passing to the event (tasks in our .js file starts from 0)

File: js/treetest.js (hunk 19 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:128]: I agree - but because it is more complicated, I think this makes the code overall less readable in this case

File: js/treetest.js (hunk 6 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js]: What's wrong with starting from 0? That's what I would expect anyway. As long as it's consistent it shouldn't really matter, right?

File: models/event.js (hunk 18 above)
Comment (rvango) [reply models/event.js]: `id` and `_id` have duplicated purpose. Use `_id` only by injecting `UUID` inside so MongoDB does not auto-generate it.

Comment (pthanvi) [reply models/event.js]: The purpose of us providing our own `id` field was to be able to uniquely identify events. Since MongoDB already supports `_id` field which does the same thing as our `id` field, it makes sense to just have one field.

Since it is not possible to get rid of `_id` field, lets just use it. However, we want to be able to provide our own value instead of MongoDB auto-generating it. MongoDB will not generate `_id` field if we provide one.

So, in short we replace `id` with `_id` :)

File: js/treetest.js (hunk 6 above)
Comment (pthanvi) [reply js/treetest.js]: Indexes starting from 0 is a programming thing. When storing in database, it makes sense to reflect how we interpret it in the real world.

File: models/event.js (hunk 18 above)
Comment (jcarstairs-scottlogic) [reply models/event.js]: Done

File: js/treetest.js (hunk 6 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js]: Done

**Review by pthanvi — APPROVED** [review]

File: js/treetest.js [hunk 21]
```diff
@@ -272,7 +272,7 @@ function bindEmitOnCloseNode() {
 }
 
 function bindEmitOnTaskChanged() {
-	window.addEventListener('taskchanged', ({ detail }) => socket.emitTaskChangedEvent(detail.newTaskIndex));
+	window.addEventListener('taskchanged', ({ detail }) => socket.emitTaskChangedEvent(detail.newTaskIndex + 1));
```
Comment (pthanvi) [inline js/treetest.js:275]: Didn't realize that other events are storing task indexes as well, and they are using zero based index. So now there is a discrepancy. 

Don't bother changing other events, lets revert this back to what it was :(

File: js/treetest.js (hunk 19 above)
Comment (rvango) [reply js/treetest.js:128]: The cognitive load is reduced because by reading the function name, you can directly assume its intent. 
If you want to look at the behaviour in isolation, you can focus on the class behaviour, but otherwise you can ignore it.

For context, on a client project, I would very likely push you to update this code. Maybe we should update it as a result. I won't enforce it this time, so it will be up-to you.

File: js/treetest.js (hunk 21 above)
Comment (jcarstairs-scottlogic) [reply js/treetest.js:275]: I've bothered changing the other events, it was just as easy
